### ISSUE TITLE
Support for structured logging in ZIO 2.x

### ISSUE THREAD

**Issue by nartamonov** [zio/zio#7000]
ZIO 2.x provides generic `ZLogger[Message, Output]` trait which allows to implement more "structured" logging where input `Message` is a product type and `Output` is JSON, for example:

```scala
abstract class LogMsg {
  def asJson: Json
}

case class AppStarted(version: String, serverHost: String) extends LogMsg {
  // Could be derived automatically by Magnolia or similar libs
  def asJson: Json =
    JsonObject(
      "msgType" -> "AppStarted",
      "version" -> version,
      "server"  -> serverHost,
      "message" -> s"Application v. $version started on server $serverHost"
    )
}

class StructuredLogger extends ZLogger[LogMsg, Json] {
  override def apply(
    trace: Trace,
    fiberId: FiberId,
    logLevel: LogLevel,
    message: () => LogMsg,
    cause: Cause[Any],
    context: FiberRefs,
    spans: List[LogSpan],
    annotations: Map[String, String]): Json = ???
}
```

I use a similar approach in my projects, using the library [Izumi Logstage](https://izumi.7mind.io/logstage/index.html). That approach has the following advantages:

1. More disciplined way to define log messages.
2. Ability to search messages in log systems that doesn't support full-text search ([Grafana Loki](https://grafana.com/oss/loki/), etc).

It would be very helpful if ZIO supports custom loggers with arbitrary `Message` and `Output` types.

Currently we have 2 obstacles to meeting this challenge:

1. We can install custom logger via `Runtime.addLogger`, but it's type have to be limited to `String => Any`.
4. We can log messages via the methods `ZIO.log*`, but the type of messages has to be `String` only.

Maybe I missed something?

**Comment by 987Nabil** [comment]
The `ZIO.log*` methods are imho not easy to change, since it would most likely break existing code nor do I think it is a good idea to have a default logging access with anything else then strings. It makes the default case to complicated.
We also output json logs, but configure this via `logback.xml`.
I don't say it is a bad idea, but I am not sure about the way.

**Comment by nartamonov** [comment]
@987Nabil agree to have only simplified `ZIO.log*` methods that works with strings only. We can make customized methods ourselves. But it should be some way to install custom loggers system-wide and to access it later.

`Logback` can format your log messages (plain strings) as JSON, but it doesn't allow you to enrich your messages by new context-specific key-value pairs. You still use string-oriented Slf4j API. So it can't replace language-side support of structured logging.

**Comment by 987Nabil** [comment]
Of course you can add context specific information, if you do somewhat like [this](https://www.baeldung.com/mdc-in-log4j-2-logback#mdc-in-slf4jlogback).
But I understand that it is more beautiful to not mess with MDC and xml, if you don't have to.

**Comment by adamgfraser** [comment]
Yes I think we can provide some functionality that lets you log something with a custom logger, maybe in `zio-logging`.

**Comment by davidlar** [comment]
FYI
We merged in json appender support to zio-logging today, that might solve your problem in another way. You can add annotations etc to the log format and it will be present in the resulting json.

```
val format = LogFormat.default + label("annotations", annotation("ann1") + annotation("ann2"))
val layer = zio.logging.consoleJson(format)
```

**Comment by hearnadam** [comment]
Closing. Please open an issue in zio-logging if you think we should extend this further.
